### ISSUE TITLE
Sampled softmax in tf keras

### ISSUE THREAD

**Issue by xinyu-Naturali** [tensorflow/tensorflow#22824]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
no
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
1.11
- **Python version**:
3.5.3
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
I want to do sampled softmax loss in tf keras.  I defined my own model by subclassing keras Model.  In __init__, I specify the layers I need including the last Dense projection layer.  But this Dense layer shouldn't be called in training as I want to do sampled softmax and only to use it's weights and biases.  Then I define the loss function like this:

    class SampledSoftmax(Layer):
        def __init__(self,
                           num_sampled,
                           num_classes,
                           projection,
                           bias,
                           hidden_size):
            self.weights = tf.transpose(projection)
            self.bias = bias
            self.num_classes = num_classes
            self.num_sampled = num_sampled
            self.hidden_size = hidden_size

        def __call__(self, y_true, input):
            """ reshaping of y_true and input to make them fit each other """
            input = tf.reshape(input, (-1,self.hidden_size))
            y_true = tf.reshape(y_true, (-1,1))

            return tf.nn.sampled_softmax_loss(
                weights=self.weights,
                biases=self.bias,
                labels=y_true,
                inputs=input,
                num_sampled=self.num_sampled,
                num_classes=self.num_classes,
                partition_strategy='div')

It takes in the necessary parameters to initialize and the class call will be the needed sampled softmax loss function.  The catch is that to add loss to model compile I need the weights etc of the last Dense.  But 1) in training Dense is not included in the model, and 2) even if it does, the Dense layer would only be hooked up with input and thus get its input dimensions etc in __call__ of my custom model.  In short, the weights etc won't be available before compiling model.  Can anyone offer some help to point me to the right direction? 

BTW, the loss defined above would work in small test cases like the following.

    x = Input(shape=(10,), name='input_x')
    emb_out = Embedding(10000,200,input_length=10)(x)
    lstm_out = LSTM(200, return_sequences=True)(emb_out)

    dense = Dense(10000, activation='sigmoid')
    output = dense(lstm_out)

    sl = SampledSoftmax(10, 10000, dense.kernel, dense.bias)

    model = Model(inputs=x, outputs=lstm_out)
    model.compile(optimizer='adam', loss=sl)
    model.summary()
    model.fit(dataset, epochs=20, steps_per_epoch=5)

**Comment by ymodak** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

**Comment by xinyu-Naturali** [comment]
I think it is a feature request.  Being able to do sampled softmax loss should be a standard feature.  When someone makes a custom model and wants to train with sampling, the loss function should share the weights with the last dense layer.  But for now there's no easy way to do it because by the time of compiling the model with the loss function, the loss function can't be defined as all the layers haven't been hooked up into a model (only happens in model call) and thus dense won't have its dimensions ready to initialize the weights.  This conflict should really be addressed

**Comment by xinyu-Naturali** [comment]
To provide an example that caused it to fail. I first subclassed model as follows:

    class LanguageModel(tf.keras.Model):
        def __init__(self, 
                 vocal_size=15003, 
                 embedding_size=512
                 input_len=64)
            self.embedding = Embedding(vocal_size, embedding_size, 
                                  input_length=input_len)
            self.lstm = LSTM(hidden_size, return_sequences=True)
            self.dense = Dense(vocal_size, activation='softmax')

        def call(self, inputs, training=False):
            emb_out = self.embedding(inputs)
            lstm_out = self.lstm(embrace_out)
            res = self.dense(lstm_out)
            if (training)
                ''' shouldn't use the last dense as we want to do sampling'''
                return lstm_out
            return res

Then the part to train the model as below

    sampled_loss = SampledLoss(num_sampled, vocal_size, 
                   model.dense.kernel, model.dense.bias,
                   hidden_size)

    model.compile(optimizer=tf.train.RMSPropOptimizer(lr),
                  loss=sampled_loss)

It would fail however I play around with it, because model.dense.kernel is not accessible as by the time of compiling the model, dense layer has not been initialized in call method.

**Comment by wt-huang** [comment]
@xinyu-Naturali Dense layer can be included in the model during training. 
Your first approach appears to be better than the second one, yet the code snippet pasted looks incomplete.
May want to check your SampledSoftmax signature, doesn't quite match.

**Comment by wt-huang** [comment]
Closing, feel free to reopen if problem persists.
